From ff1290fb1452b183396378de95ab200afc46ea04 Mon Sep 17 00:00:00 2001
From: Jinqian Yang <yangjinqian1@huawei.com>
Date: Fri, 29 Aug 2025 10:58:36 +0800
Subject: [PATCH 774/960] downstream: KVM: arm64: fix the issue of vNMI not
 supporting live migration

Add the migration of GICR_INMIR0. vNMI migration for SPI is not
supported, only supports SGI and PPI in guest.

Signed-off-by: Jinqian Yang <yangjinqian1@huawei.com>
Signed-off-by: Xie Xiaodong <xiexiaodong16@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kvm/vgic/vgic-mmio-v3.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/kvm/vgic/vgic-mmio-v3.c b/arch/arm64/kvm/vgic/vgic-mmio-v3.c
index 4db39e9f7372..eacd44764f6f 100644
--- a/arch/arm64/kvm/vgic/vgic-mmio-v3.c
+++ b/arch/arm64/kvm/vgic/vgic-mmio-v3.c
@@ -643,6 +643,12 @@ static unsigned long vgic_mmio_read_nmi(struct kvm_vcpu *vcpu,
 	return value;
 }
 
+static unsigned long vgic_uaccess_read_nmi(struct kvm_vcpu *vcpu,
+					gpa_t addr, unsigned int len)
+{
+	return vgic_mmio_read_nmi(vcpu, addr, len);
+}
+
 static void vgic_mmio_write_nmi(struct kvm_vcpu *vcpu, gpa_t addr,
 				unsigned int len, unsigned long val)
 {
@@ -672,6 +678,13 @@ static void vgic_mmio_write_nmi(struct kvm_vcpu *vcpu, gpa_t addr,
 	}
 }
 
+static int vgic_uaccess_write_nmi(struct kvm_vcpu *vcpu, gpa_t addr,
+				unsigned int len, unsigned long val)
+{
+	vgic_mmio_write_nmi(vcpu, addr, len, val);
+	return 0;
+}
+
 /*
  * The GICv3 per-IRQ registers are split to control PPIs and SGIs in the
  * redistributors, while SPIs are covered by registers in the distributor
@@ -832,8 +845,9 @@ static const struct vgic_register_region vgic_v3_rd_registers[] = {
 	REGISTER_DESC_WITH_LENGTH(SZ_64K + GICR_NSACR,
 		vgic_mmio_read_raz, vgic_mmio_write_wi, 4,
 		VGIC_ACCESS_32bit),
-	REGISTER_DESC_WITH_LENGTH(SZ_64K + GICR_INMIR0,
-		vgic_mmio_read_nmi, vgic_mmio_write_nmi, 4,
+	REGISTER_DESC_WITH_LENGTH_UACCESS(SZ_64K + GICR_INMIR0,
+		vgic_mmio_read_nmi, vgic_mmio_write_nmi,
+		vgic_uaccess_read_nmi, vgic_uaccess_write_nmi, 4,
 		VGIC_ACCESS_32bit),
 };
 
-- 
2.43.0

